Userland/Services/SystemServer/Service.cpp

```diff
@@ -144,14 +157,38 @@ void Service::handle_socket_connection()
 
 void Service::activate()
 {
-    VERIFY(m_pid < 0);
+    // Already running
+    if (m_pid >= 0)
+        return;
+    
+    // It makes that the service won't restart automatically.
+    m_force_stopped = false;
 
     if (m_lazy)
         setup_notifier();
     else
         spawn();
 }
 
+void Service::deactivate()
+{
+    // Already stopped
+    if (m_pid <= 0)
+        return;
+    
+    // It makes that the service won't restart automatically.
+    m_force_stopped = true;
+
+    if (m_lazy)
+        m_socket_notifier = nullptr;
+    else {
+        // FIXME: Maybe add some configuration option for it ('AllowDeactivate')
+        // FIXME: Check for errors
+        kill(m_pid, SIGTERM);
+        did_exit(0);
```

Comment:
No no no, you don't call `did_exit()` yourself, it gets called when you receive a notification from the kernel that it did, in fact, exit — via SIGCHLD, or potentially via other means.

Also this is not how you kill services. You kill with SIGTERM, yes, then you wait for some timeout and then you kill with SIGKILL. And importantly, you make super sure this does not race with you reaping the child, so that you never kill another process that happens to reuse the same PID. This is not a trivial task.

Reply:
> No no no, you don't call `did_exit()` yourself, it gets called when you receive a notification from the kernel that it did, in fact, exit — via SIGCHLD, or potentially via other means.

This wasn't called then for some reason.

Reply:
Should it be something like this?
```c++
Service::Service()
{
    ...
    m_sigterm_timer = Core::Timer::create_single_shot(5000, [this]() {
        if (m_pid == m_last_pid) {
            dbgln("Force killing service {} ({})", name(), m_pid);
            kill(m_pid, SIGKILL);
        }
    });
    ....
}
...
void Service::deactivate()
{
    ....
    m_sigterm_timer->start();
    kill(m_pid, SIGTERM);
    ....
}
```
(`m_last_pid` is set to `m_pid` in `did_exit` just before `m_pid` is set to -1)

Reply:
`m_last_pid` doesn't sound too good. Can't we simply disarm the timer if the service exits before the time runs out?

Also, you have to check whether the `kill()` calls in fact succeed, and at least log an error otherwise.

Reply:
Simply
```
m_sigterm_timer->stop();
m_pid = -1;
```
in did_exit and
```
kill(m_pid, SIGKILL);
```
in m_sigterm_timer?

Reply:
Yeah I guess.

But again, please make very sure that killing a process and reaping it cannot race. That mainly means we should never reap from inside the `SIGCHLD` handler, only inside the event loop.